### ISSUE TITLE
Allow label_join to be used in aggregation operators

### ISSUE THREAD

**Issue by PrestonTaylor** [prometheus/prometheus#9850]
I have looked quite a bit and haven't found a way to do this but it seems simple so please let me know if I've missed something.

I have a need to do capacity planning for failover events which rarely happen. So I can't rely on the data to cover these scenarios. However; assuming a label that designates a datacenter for example I can aggregate while maintaining these label values for further manipulation in outer queries which would eliminate the need to do a bunch of manipulation in other places like excel, grafana etc. This is necessary in my situation because my failover scenarios are extremely complex and just calculating totals isn't helpful.

	{instance="datacenter0"}
	1000
	{instance="datacenter1"}
	2000
	
	sum by(label_join("instance", "datacenters" ",")) (
		irate(
			tIPFilterParamsEgrHitByteCount{
				instance=~"datacenter.*", 
			}[10m]
		) 
	)
	
	{datacenters="datacenter0,datacenter1"}
	3000

**Comment by bboreham** [comment]
You want a new label which contains all the values of a label that disapeared under aggregation?

Interesting idea, though I don't think `label_join` is the right name for that.

**Comment by PrestonTaylor** [comment]
Correct, similar to GROUP_CONCAT() in MySQL. It would be helpful to maintain those for presentation layer as well as further manipulation in a regex or things like that.

Sure, I was just using that as an example, perhaps group_concat might work as a function name?

**Comment by bboreham** [comment]
It wouldn't be a function, in the sense that functions work currently.
It could be added as an optional part of `sum`, maybe `sum by(instance) concat_as(datacenters) (...)`.
Maybe some cryptic code like `%` as was done for evaluation timestamp.

Anyway, my intention was not to nitpick the name but to check if I understood the suggestion.

**Comment by JessicaGreben** [comment]
@SuperQ / @juliusv  We looked at this issue during the bug scrub meeting and we don't think this should be part of the language, but we want to get more input from you.

**Comment by juliusv** [comment]
It's an interesting feature request which I never saw requested before, but which doesn't seem too unreasonable either. Such a modifier to aggregators would easily fit into the language model (you'd probably want to be able to specify the concatenation separator as an optional parameter, similar to relabeling rules), but personally I'd want to see more interest from other people before considering adding it to the language.

Another question would be how to syntactically indicate which aggregated label names get concatenated into which output labels, since a simple `sum(foo)` could aggregate over arbitrarily many label dimensions, and you may want to preserve some of them as concatenated lists and others not.

**Comment by ostrya** [comment]
I have a similar use case and would also like to have a feature like this. In my case, I have several instances of an application, and I monitor based on the up{application="foo"} metric. Now I would like to create an alert query that indicates if at least two instances are down (up == 0), and ideally the resulting time series should contain the list of instances which are down.

The best I can currently do is the following, but that results in two separate time series:

```
(sum(clamp_min(up{application="foo"} == 0, 1)) >= 2) + ignoring(instance) group_right() (sum by (instance) (up{application="foo"} == 0))

-> {instance="instance1"}, {instance="instance2"}
```

But with the ability to concatenate labels, I would be able to get one metric which contains both instance names, e.g. like this:

```
(sum(clamp_min(up{application="foo"} == 0, 1)) >= 2) + ignoring(instances) group_right() (sum by (instance) concat_as(instances) (up{application="foo"} == 0))

-> {instances="instance1,instance2"}
```

**Comment by paulolieuthier** [comment]
I'm also looking forward to this feature. I think it applies to all scenarios where one wants an alert on an aggregation but wants the alert to carry the list of offenders (through the label).

**Comment by rwultsch** [comment]
This feature would be very helpful.

**Comment by juliusv** [comment]
Interesting, ok! By now a few more people have voiced interest in this feature here, and I think it's at least worth considering if we get a design proposal that takes into account the following points:

* How to specify which labels you want to preserve as a concatentation.
* How to set the separator character (defaulting to ',' or something when omitted).
* How to specify whether you want the concatenated values to be distinct rather than repeated (some underlying label values in an aggregation can occur multiple times). The analogue to this in MySQL would be `GROUP_CONCAT(DISTINCT ...)`.
* Optional(?): How to specify the sorting of the concatenated labels. Maybe just always keeping this lexicographic is fine though.

Am I missing anything?

If we want to add these extra modifiers for separators and value distinctness at least, then coming up with a syntax that isn't too complex and still PromQL-ish is going to require a bit of creativity. Ideas welcome!

**Comment by beorn7** [comment]
Hello from the bug scrub!

Things became a bit quiet here, but we decided to not close this yet. As @juliusv said, this needs good ideas to solve it without creating an overly complex syntax.

**Comment by Nikokolas3270** [comment]
Hello,

Just some +1 on this feature request, I would also benefit from the proposed label concatenation while aggregating during my day to day work.
Concerning the proposed syntax, what about something simple like:
```
<aggregator> by (<label list>) concat(dst_label, src_label [, separator]) (<promql expr>)
```
?
Remark that the `separator` would be optional and default to `","`

I can take a look at the coding if you want

**Comment by juliusv** [comment]
@Nikokolas3270 The `dst_label` / `src_label` in your proposal suggests that you're assuming only a single label being aggregated away and needing concatenation? I think this would have to support multiple labels being aggregated away (anything that is not mentioned in the `by()` clause, or anything that *is* mentioned in the `without()` clause). But it could always concatenate into the original label name, so `dst_label` and `src_label` might not be needed.

The simplest form would be just a `concat()` clause that would take an optional separator that defaults to e.g. `,`:

Input data:

```
my_metric{path="/api/foo", method="POST", status="200"}
my_metric{path="/api/foo", method="POST", status="500"}
my_metric{path="/api/foo", method="GET", status="200"}
my_metric{path="/api/foo", method="GET", status="500"}
my_metric{path="/api/bar", method="POST", status="200"}
my_metric{path="/api/bar", method="POST", status="404"}
my_metric{path="/api/bar", method="GET", status="200"}
my_metric{path="/api/bar", method="GET", status="500"}
```

Expression:

```
sum by(path) concat() (my_metric)
```

Output:

```
{path="/api/foo", method="GET,POST", status="200,500"}
{path="/api/bar", method="GET,POST", status="200,404,500"}
```

Or with an explicit separator:

```
sum by(path) concat(";") (my_metric)
```

Output:

```
{path="/api/foo", method="GET,POST", status="200;500"}
{path="/api/bar", method="GET,POST", status="200;404;500"}
```

Of course it would be nice to give more control (sorting, distinctness, label renaming, etc.) and also allow specifying that separately for each individual label that is being aggregated away, but then it gets hairy syntactically with multiple options for multiple labels each. I guess one could have multiple `concat()` clauses and have them mention the label name, along with optional options. Like `sum by(foo) concat(datacenters) concat(methods, ";") (bar)`.

Just thinking out loud here. I don't know that I'm the right person to come up with a good design here, or make the judgement call.
